### PR TITLE
Implement minimally-viable HTTP request & response serialization

### PR DESCRIPTION
**Pull request by bgerstle**
> Necessary for #15

Implements basic `Request` and `Response` value types, which can be read/written to/from streams. There are some edge cases to cover and I'd like to add some more (ideally parameterized) tests to cover these basics before moving into headers and payloads (aka bodies).

### CODE REVIEW

**Comment by trptcolin** [comment]
Cool that you're exploring property-based testing. That stuff looks good and I see you already took care of some of the missing unit test coverage for `Request` - great. 👍 

Might be interesting, as a guard rail, to generate test coverage reports and add mutation testing to your toolkit. I don't have strong opinions on which tools, but mutation testing is kind of fun - it can be a nice way to help ensure that breaking production code will break tests too.

File: src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+public class HttpHEADAcceptanceTest extends TcpServerIntegrationTest {
+  @BeforeEach
+  void startAndConnect() throws IOException {
+    server.start();
+    client.connect();
+  }
+
+  @Test
+  void simpleHEADRequest() throws Throwable {
+    Request req = new Request("1.1", Method.HEAD, new URI("/test")) ;
```
[the remark this answers is not in the document]
Comment (bgerstle) [reply src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java]: Interesting, I'll try to repro. Otherwise, I still stand by my use of acceptance tests here, as it's way faster than me running `curl` on the command line, modifying the code, re-running the app, etc. Of course, I'm still in the process of adding more unit tests 😉

**Comment by bgerstle** [comment]
I had also started wondering about the code coverage, but wanted to focus more on writing tests & refactoring first—and I'm wary of falling down the rabbit hole of getting it working in Intellij, etc...

Mutation testing sounds very interesting, I'll try to look into it if I have time.

File: src/main/java/com/eighthlight/fabrial/http/Response.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(version, statusCode, reason);
+  }
+
+  @Override
+  public String toString() {
+    return "Response{" +
+           "statusCode=" + statusCode +
```
[the remark this answers is not in the document]
Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/http/Response.java]: Yet another reason to scrap these hand-rolled value types and go with something that will generate the classes for me...

File: src/test/java/com/eighthlight/fabrial/test/http/HttpRequestLineParsingTests.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  void getWithoutHeaders()
+      throws URISyntaxException, RequestParsingException, IOException {
+    URI uri = new URI("/");
+    InputStream is = new ByteArrayInputStream("GET / HTTP/1.1\r\n".getBytes(StandardCharsets.UTF_8));
+    assertThat(Request.readFrom(is),
+               equalTo(new Request("1.1", Method.GET, new URI("/"))));
+  }
+
+  @Test
+  void leadingWhitespace() {
```
[the remark this answers is not in the document]
Comment (bgerstle) [reply src/test/java/com/eighthlight/fabrial/test/http/HttpRequestLineParsingTests.java]: Good to know about `@Disabled` 👍

File: src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (trptcolin) [reply src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java]: Well sure, manual tests << integration/acceptance/e2e tests << unit tests.

[the remark this answers is not in the document]
Comment (trptcolin) [reply src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java]: Well sure, manual tests << integration/acceptance/e2e tests << unit tests.

**Comment by trptcolin** [comment]
Yeah, low priority for mutation testing / code coverage.

File: src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bgerstle) [reply src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java]: Acceptance tests passing locally for me: `./gradlew testAcceptance -i`. I'm guessing that it's either an issue with the port being used on your machine, or (IME, infrequent) race conditions between the server starting and the client's connection attempt. If it proves to be flakey in CI as well, I'll DRY up the setup logic so it retries the client connection before running tests.

[the remark this answers is not in the document]
Comment (bgerstle) [reply src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java]: Sorry, this is actually an integration test (rename commit incoming). The whole suite is passing locally for me. There are some error logs written during the test, but as long as it shows "passing" it should be fine.

[the remark this answers is not in the document]
Comment (trptcolin) [reply src/test/java/com/eighthlight/fabrial/test/server/HttpHEADAcceptanceTest.java]: I'm seeing the issue when running it via IntelliJ "all tests", depending on the ordering, and I am now seeing that it's intermittent. 

As a smaller, more focused case, if I highlight those two test classes together (`AppAcceptanceTest` and `HttpHeadIntegrationTest`), and run those (in that order), the second one sometimes fails with the error I mentioned. Perhaps that will allow you to reproduce?

I do see the suite passing when I run the separate suites that CI runs, for what it's worth - in this case those tests never run quickly enough back-to-back to uncover the shared resource conflict.

When running tests in the IDE, running "all tests" tends to be a default for a lot of folks, so even if it passes in CI, I think this is worth addressing. One practice I've seen that would have solved the *direct* bug for me would be to use this separate `test-integration` source directory that you have in your `build.gradle` (but I think aren't using?) for slow tests. This wouldn't solve the underlying root cause (competition for a single shared resource, the port), but would make it less likely to encounter.

**Comment by bgerstle** [comment]
So, the `Request` class has now been heavily refactored to break up the parsing of individual fields (using the Builder pattern). Next, I plan to refactor the `HttpConnectionHandler` to extract the request/response logic so it can be tested using `Request`/`Response` objects instead of streams & string equality checks... THEN I'll finally be able to add in a file directory routing and get some real functionality going.
